Add option to use codeception

poohnix opened the pull request:
Problem:
While PHPUnit is still very popular many projects now use Codeception (which includes PHPUnit). The syntax is slightly different for some options so out of the box it is not possible to run all better-phpunit commands with Codeception.

Solution:
Add a new option to specify that you want to use Codeception instead of PHPUnit. With the option "useCodeception" (which is false by default) the syntax is changed accordingly so that you are able to use Codeception (provided that you specified it's binary in the option "better-phpunit.phpunitBinary".

Remark:
This patch obviously had to include my previous one for fixing the extension tests.
Attention: For brevity I have also included my other patch (runSuites) in this one!

calebporzio commented:
Thanks for putting so much time and effort into the project. It's so nice when people open PR's instead of issues lol.

I'm open to supporting codeception, but would want to hear that more people are looking for it before I pull it in (especially because I don't use it and wouldn't be able to support it).

poohnix commented:
Hi Caleb,

with the broader option runSuiteSuffix (see latest commit in PR #84) it is now possible to back out the changes of #84 completely as there is no difference any more between PHPUnit and Codeception. Just let me know if you would feel more comfortable about this PR if I would remove the changes of PR #84 from this branch.

Thank you,
Hank

calebporzio commented:
Couple things:
1) Thank you for clearly laying out the similarities between running codeception and phpunit - that does ease my fear
2) The idea of it being the defacto Codeception extension in the marketplace IS enticing. However, its title still has "PHPUnit" in it, and not "Codeception" so I don't imagine people will find it searching. Also, IF it  did rise to popularity for that tool, it wouldn't be as focused as it is now, and might be confusing for PHPUnit users. I could see us handling it in such a way that it's less emphasized in the docs, but I would want to make sure it doesn't get too "BIG" and "optiony" fealing for PHPUnit users.
3) We already have the "better-phpunit.commandSuffix" option. how is runSuiteSuffix different? Seems confusing
4) Sorry I'm not keeping better track, but can you make sure all your PR's are isolated and don't depend on each other (which I think is what you did, just want to make sure before I totally wrap my mind around everything)
5) Is there a way you could achieve codeception support just by utilizing the existing options? Or by adding a generic option or two? I'd much rather do something like that than make support a first-class citizen.

Thanks!

poohnix commented:
Thanks for getting back to me, I'll try to briefly answer your questions (sorry for getting carried away and that lengthy in my other answers).

1. You are welcome and indeed the changes needed to support Codeception are rather small.
2. I found your extension while searching for a Codeception extension because knowing Codeception is based on PHPUnit this was the second search term I looked for after not finding anything Codeception related. ;-) And as this is only one more option I don't think it will get too confusing for PHPUnit users (as it is off by default anyway) - but obviously I'm rather biased. ;-)
3. While I changed the option in the other PR to this name I was already thinking about this, because I would have normally named the option rather lengthy "additionalSuiteOptions" but wanted to stay in line with your current naming. The difference is that these options are only used with the "run suite" command as for example it hardly makes any sense to specify "--coverage" as a global suffix which would be run for a single file or even a single method.
4. You are totally right I shouldn't have mixed this PR with #84. I normally don't do this but in this case by the time I did it it would have meant much more work keeping them separate. Now that I changed the structure of #84 this is no longer the case. So before you invest any more time in these two PR I would rather redo my work in two separate PR now that I've streamlined the changes a bit more.
5. I have no problem with calling the option "useOtherTestTool" or similar (also I'm not sure how many others there actually are for PHP apart from PHPUnit and Codeception). With the existing options it was not possible to run Codeception because it uses a different syntax for specifying a single test method and it doesn't use a phpunit.xml file (the other differences could be "emulated" by using "phpunitBinary": "/usr/local/bin/codecept run").

Cheers!

poohnix commented:
If this PR is of interest to you please have a look at PR #91 which is a streamlined and isolated version of the above changes and should be way easier to comprehend.
